app/routes.py

```diff
@@ -1,2 +1,193 @@
+    if not task.check_if_completed():
+        slack_api_url = "https://slack.com/api/chat.postMessage"
+        headers = {"Authorization": "Bearer " + os.environ.get("SLACK_API_KEY")}
+        param_payload = {
+            "channel": "task-notifications", 
+            "text": f"Someone has just completed the task {task.title}"
+            }
+        
+        try:
+            requests.post(slack_api_url, headers=headers, params=param_payload)
+        
+        except Exception as e:
+            return f"Error posting message to Slack: {e}"
+
+    # Change task to completed in db
+    task.completed_at = datetime.now()
+    db.session.commit()
+
+    return {"task": task.to_dict()}, 200
+
+@tasks_bp.route("<task_id>/mark_incomplete", methods=["PATCH"])
+@require_instance_or_404
+def update_task_to_incomplete(task):
+    """Updates task at particular id to incomplete using PATCH."""
+    task.completed_at = None
+    db.session.commit()
+    return {"task": task.to_dict()}, 200
+
+@tasks_bp.route("/<task_id>", methods=["DELETE"])
+@require_instance_or_404
+def delete_task(task):
+    """Deletes task by id."""
+    db.session.delete(task)
+    db.session.commit()
+
+    return {
+        "details": f"Task {task.id} \"{task.title}\" successfully deleted"
+    }, 200
+
+@goals_bp.route("", methods=["GET"])
+def get_goals():
+    """Retrieve all stored goals."""
+    goals = Goal.query.all()
+
+    return jsonify([goal.to_dict() for goal in goals]), 200
+
+@goals_bp.route("/<goal_id>", methods=["GET"])
+@require_instance_or_404
+def get_goal(goal):
+    """Retrieve one stored goal by id."""
+    return jsonify({"goal": goal.to_dict()}), 200
+
+@goals_bp.route("", methods=["POST"])
+def create_goal():
+    """Create a new goal from JSON data."""
+    form_data = request.get_json()
+
+    if "title" not in form_data:
+        return jsonify({"details": "Invalid data"}), 400
+
+    new_goal = Goal(
+        title=form_data["title"]
+    )
+    db.session.add(new_goal)
+    db.session.commit()
+
+    return jsonify({"goal": new_goal.to_dict()}), 201
+
+@goals_bp.route("/<goal_id>", methods=["PUT"])
+@require_instance_or_404
+def update_goal(goal):
+    """Updates goal by id."""
+    form_data = request.get_json()
+
+    goal.update_from_dict(form_data)
+    db.session.commit()
+
+    return jsonify({"goal": goal.to_dict()}), 200
+
+@goals_bp.route("/<goal_id>", methods=["DELETE"])
+@require_instance_or_404
+def delete_goal(goal):
+    """Deletes goal by id."""
+    db.session.delete(goal)
+    db.session.commit()
+
+    return {
+        "details": f"Goal {goal.id} \"{goal.title}\" successfully deleted"
+    }, 200
+
+@goals_bp.route("/<goal_id>/tasks", methods=["POST"])
+@require_instance_or_404
+def post_tasks_related_to_goal(goal):
+    """Adds tasks to goal wiht id."""
+    form_data = request.get_json()
+
+    for task_id in form_data["task_ids"]:
+        query = Task.query.get(task_id)
+        if not query:
```

Comment:
Very minor suggestion, `query` is a little confusing here as a variable name as the value being stored is a task not a query, I'd recommend something like `task` or `current_task`.